internal/util/oem/azure/report.go

```diff
@@ -0,0 +1,253 @@
+	ExpectHealthReport    string                  `xml:"Machine>ExpectHealthReport"`
+	ContainerId           string                  `xml:"Container>ContainerId"`
+	RoleInstanceList      []GoalStateRoleInstance `xml:"Container>RoleInstanceList>RoleInstance"`
+}
+
+type GoalStateRoleInstance struct {
+	XMLName                  xml.Name `xml:"RoleInstance"`
+	InstanceId               string   `xml:"InstanceId"`
+	State                    string   `xml:"State"`
+	HostingEnvironmentConfig string   `xml:"Configuration>HostingEnvironmentConfig"`
+	SharedConfig             string   `xml:"Configuration>SharedConfig"`
+	ExtensionsConfig         string   `xml:"Configuration>ExtensionsConfig"`
+	FullConfig               string   `xml:"Configuration>FullConfig"`
+	Certificates             string   `xml:"Configuration>Certificates"`
+	ConfigName               string   `xml:"Configuration>ConfigName"`
+}
+
+// getClient returns an HTTP client wrapper that will perform retries when
+// performing HTTP requests. Required headers for interacting with the Azure
+// servers will be set on all requests.
+func getClient() retry.Client {
+	client := retry.Client{
+		InitialBackoff: time.Second,
+		MaxBackoff:     time.Second * 5,
+		MaxAttempts:    10,
+		Header: map[string][]string{
+			"x-ms-agent-name": {agentName},
+			"x-ms-version":    {fabricProtocolVersion},
+			"Content-Type":    {"text/xml; charset=utf-8"},
+		},
+	}
+
+	return client
+}
+
+// buildHealthReport will convert the goal state and the desired message into an
+// XML-encoded health report message
+func buildHealthReport(incarnation, container_id, role_instance_id, status, substatus, description string) ([]byte, error) {
+	var details *HealthDetails
+	if substatus != "" {
+		details = &HealthDetails{
+			SubStatus:   substatus,
+			Description: description,
+		}
+	}
+	data, err := xml.MarshalIndent(HealthReport{
+		Xsi:                  "http://www.w3.org/2001/XMLSchema-instance",
+		Xsd:                  "http://www.w3.org/2001/XMLSchema",
+		GoalStateIncarnation: incarnation,
+		ContainerId:          container_id,
+		InstanceId:           role_instance_id,
+		State:                status,
+		Details:              details,
+	}, "  ", "    ")
+	if err != nil {
+		data = append([]byte(xml.Header), data...)
+	}
+	return data, err
+}
+
+// getGoalState will send a request to the given IP address for the XML-encoded
+// goal state of this machine, and then decodes the message and returns it.
+func getGoalState(addr net.IP) (GoalState, error) {
+	xmlBlob, err := getClient().Getf(GoalStateUri, addr.String())
+	if err != nil {
+		return GoalState{}, fmt.Errorf("failed to fetch goal state: %v", err)
+	}
+	var gs GoalState
+	err = xml.Unmarshal(xmlBlob, &gs)
+	if err != nil {
+		return GoalState{}, err
+	}
+	return gs, nil
+}
+
+// findLease will attempt to find a systemd lease file for one of the interfaces
+// enumerated by net.Interfaces()
+func findLease() (*os.File, error) {
```

Comment:
We could move some of the azure logic in coreos-metadata into a top level `pkg/` package or something, but I'm not sure about the implications of adding a dependency from one to the other. Unless we want to make a new `libazure` library.